src/Simulation/qdk_sim_rs/src/c_api.rs

```diff
@@ -0,0 +1,443 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// The following two attributes include the README.md for this module when
+// building docs (requires +nightly).
+// See https://github.com/rust-lang/rust/issues/82768#issuecomment-803935643
+// for discussion.
+#![cfg_attr(doc, feature(extended_key_value_attributes))]
+#![cfg_attr(doc, cfg_attr(doc, doc = include_str!("../docs/c-api.md")))]
+
+use crate::{built_info, NoiseModel, Process, State};
+use lazy_static::lazy_static;
+use serde_json::json;
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::ptr;
+use std::sync::Mutex;
+
+struct CApiState {
+    register_state: State,
+    noise_model: NoiseModel,
+}
+
+lazy_static! {
+    static ref STATE: Mutex<HashMap<usize, CApiState>> = Mutex::new(HashMap::new());
+    static ref LAST_ERROR: Mutex<Option<String>> = Mutex::new(None);
+}
+
+// UTILITY FUNCTIONS //
+
+/// Exposes a result to C callers by setting LAST_ERROR in the Error
+/// case, and generating an appropriate error code.
+fn as_capi_err<F: FnOnce() -> Result<(), String>>(result_fn: F) -> i64 {
+    let result = result_fn();
+    match result {
+        Ok(_) => 0,
+        Err(msg) => {
+            *LAST_ERROR.lock().unwrap() = Some(msg);
+            -1
+        }
+    }
+}
+
+fn apply<F: Fn(&NoiseModel) -> &Process>(
+    sim_id: usize,
+    idxs: &[usize],
+    channel_fn: F,
+) -> Result<(), String> {
+    let state = &mut *STATE.lock().unwrap();
+    if let Some(sim_state) = state.get_mut(&sim_id) {
+        let channel = channel_fn(&sim_state.noise_model);
+        match channel.apply_to(idxs, &sim_state.register_state) {
+            Ok(new_state) => {
+                sim_state.register_state = new_state;
+                Ok(())
+            }
+            Err(err) => Err(err),
+        }
+    } else {
+        return Err(format!("No simulator with id {}.", sim_id));
+    }
+}
+
+// C API FUNCTIONS //
+
+/// Returns information about how this simulator was built, serialized as a
+/// JSON object.
+#[no_mangle]
+pub extern "C" fn get_simulator_info() -> *const c_char {
+    let build_info = json!({
+        "name": "Microsoft.Quantum.Experimental.Simulators",
+        "version": built_info::PKG_VERSION,
+        "opt_level": built_info::OPT_LEVEL,
+        "features": built_info::FEATURES,
+        "target": built_info::TARGET
+    });
+    CString::new(serde_json::to_string(&build_info).unwrap().as_str())
+        .unwrap()
+        .into_raw()
+}
+
+/// Returns the last error message raised by a call to a C function.
+/// If no error message has been raised, returns a null pointer.
+#[no_mangle]
+pub extern "C" fn lasterr() -> *const c_char {
+    match &*LAST_ERROR.lock().unwrap() {
+        None => ptr::null(),
+        Some(msg) => CString::new(msg.as_str()).unwrap().into_raw(),
+    }
+}
```

Comment:
(Not necessarily in this PR, just something to keep in mind, _if applicable_. I don't know Rust)

I have an impression that `lasterr()` returns a pointer to the _heap-allocated block_. If that's the case then the users are likely to expect a mechanism (in the same binary that made an allocation) to deallocate such blocks (and documentation about that).